cmds/jfrogplugin/uploaders/helm/helm.go

```diff
@@ -0,0 +1,248 @@
+	purl, err := parseURLAllowNoScheme(spec.URL)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse URL: %w", err)
+	}
+
+	var info ppi.UploadTargetSpecInfo
+
+	// By default, we identify an artifactory repository as a combination
+	// of Host & Repository
+	info.ConsumerId = credentials.ConsumerIdentity{
+		cpi.ID_TYPE:   NAME,
+		ID_HOSTNAME:   purl.Hostname(),
+		ID_REPOSITORY: spec.Repository,
+	}
+	if purl.Port() != "" {
+		info.ConsumerId.SetNonEmptyValue(ID_PORT, purl.Port())
+	}
+
+	return &info, nil
+}
+
+// EnsureSpecWithHelpFromHint introspects the hint and fills the target spec based on it.
+// It makes sure that the spec can be used to access a JFrog Artifactory HELM Repository.
+func EnsureSpecWithHelpFromHint(spec *JFrogHelmUploaderSpec, hint string) error {
+	if refFromHint, err := reference.Parse(hint); err == nil {
+		if refFromHint.Digest() != "" && refFromHint.Object == "" {
+			return fmt.Errorf("the hint contained a valid reference but it was a digest, so it cannot be used to deduce a version of the helm chart: %s", refFromHint)
+		}
+		if spec.ChartName == "" {
+			spec.ChartVersion = refFromHint.Object
+		}
+		if spec.ChartName == "" {
+			spec.ChartName = path.Base(refFromHint.Locator)
+		}
+	}
+	if spec.ChartName == "" {
+		return fmt.Errorf("the chart name could not be deduced from the hint (%s) or the config (%s)", hint, spec)
+	}
+	if spec.ChartVersion == "" {
+		return fmt.Errorf("the chart version could not be deduced from the hint (%s) or the config (%s)", hint, spec)
+	}
+	return nil
+}
+
+// ConvertTargetSpecToHelmUploadURL interprets the JFrogHelmUploaderSpec into a valid REST API Endpoint URL to upload to.
+// It requires a valid ChartName and ChartVersion to determine the correct URL endpoint.
+//
+//	See https://jfrog.com/help/r/jfrog-rest-apis/deploy-artifact for the URL endpoint
+//	See https://jfrog.com/help/r/jfrog-artifactory-documentation/deploying-artifacts for artifact deployment reference
+//	See https://jfrog.com/help/r/jfrog-artifactory-documentation/use-the-jfrog-helm-client for the HELM Client reference.
+//
+// Example:
+//
+//	JFrogHelmUploaderSpec.URL => demo.jfrog.ocm.software
+//	JFrogHelmUploaderSpec.Repository => my-charts
+//	JFrogHelmUploaderSpec.ChartName => podinfo
+//	JFrogHelmUploaderSpec.ChartVersion => 0.0.1
+//
+// will result in
+//
+//	url.URL => https://demo.jfrog.ocm.software/artifactory/my-charts/podinfo-0.0.1.tgz
+func ConvertTargetSpecToHelmUploadURL(spec *JFrogHelmUploaderSpec) (*url.URL, error) {
+	requestURL := path.Join(spec.URL, "artifactory", spec.Repository, fmt.Sprintf("%s-%s.tgz", spec.ChartName, spec.ChartVersion))
+	requestURLParsed, err := parseURLAllowNoScheme(requestURL)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse full request URL: %w", err)
+	}
+	return requestURLParsed, nil
+}
+
+// parseURLAllowNoScheme is an adaptation / hack on url.Parse because
+// url.Parse does not support parsing a URL without a prefixed scheme.
```

Comment:
What? It totally does.

```go
	parsedURL, err := url.Parse("github.com")
	if err != nil {
		return nil, err
	}

        // out -> github.com
	fmt.Println(parsedURL)
```

Reply:
Technically youre correct that it does follow that on all strict semantic urls, however host:port combinations that are not a URL do not work. Example:

```go
package main

import (
	"fmt"
	"net/url"
	"os"
)

func main() {
	for _, tc := range []string {
		"github.com",
		"//www.google.com/support/websearch/?source=g&amp;hl=en",
		"localhost",
		"127.0.0.1",
		"127.0.0.1:8080",
		"dummy://127.0.0.1:8080",
		"localhost:8080",
		"dummy://localhost:8080",
		"https://demo.jfrog.ocm.software/artifactory/my-charts/podinfo-0.0.1.tgz",
		"demo.jfrog.ocm.software/artifactory/my-charts/podinfo-0.0.1.tgz",
	} {
		url, err := url.Parse(tc)
		if err != nil {
			fmt.Fprintln(os.Stderr, err.Error())
			continue
		}
		fmt.Printf("URL: %s\n Scheme: %s\n", url, url.Scheme)
	}
}
```

results in 
```
parse "127.0.0.1:8080": first path segment in URL cannot contain colon
URL: github.com
 Scheme: 
URL: //www.google.com/support/websearch/?source=g&amp;hl=en
 Scheme: 
URL: localhost
 Scheme: 
URL: 127.0.0.1
 Scheme: 
URL: dummy://127.0.0.1:8080
 Scheme: dummy
URL: localhost:8080
 Scheme: localhost
URL: dummy://localhost:8080
 Scheme: dummy
URL: https://demo.jfrog.ocm.software/artifactory/my-charts/podinfo-0.0.1.tgz
 Scheme: https
URL: demo.jfrog.ocm.software/artifactory/my-charts/podinfo-0.0.1.tgz
 Scheme: 
```

as you can see, some ip combinations as well as local host are not technically URLs so they will be incorrectly identified. As such the scheme defaulting might fail, which is why this code exists. If you think those cases are not relevant im the first to vote for removal of that code